WalletWasabi/Blockchain/BlockFilters/IndexBuilderService.cs

```diff
@@ -118,148 +90,105 @@ public void Synchronize()
 			{
 				try
 				{
-					if (Interlocked.Read(ref _runner) >= 2)
+					if (Interlocked.Read(ref _workerCount) >= 2)
 					{
 						return;
 					}
 
-					Interlocked.Increment(ref _runner);
-					while (Interlocked.Read(ref _runner) != 1)
+					Interlocked.Increment(ref _workerCount);
+					while (Interlocked.Read(ref _workerCount) != 1)
 					{
 						await Task.Delay(100);
 					}
 
-					if (Interlocked.Read(ref _running) >= 2)
+					if (Interlocked.Read(ref _serviceStatus) >= Stopping)
 					{
 						return;
 					}
 
 					try
 					{
-						Interlocked.Exchange(ref _running, 1);
+						Interlocked.Exchange(ref _serviceStatus, Running);
 
-						var isImmature = false; // The last 100 blocks are reorgable. (Assume it is mature at first.)
 						SyncInfo syncInfo = null;
 						while (IsRunning)
 						{
 							try
 							{
 								// If we did not yet initialized syncInfo, do so.
-								if (syncInfo is null)
+								if (syncInfo == null)
 								{
 									syncInfo = await GetSyncInfoAsync();
 								}
 
-								uint heightToRequest = StartingHeight;
+								uint currentHeight = 0;
 								uint256 currentHash = null;
 								using (await IndexLock.LockAsync())
 								{
 									if (Index.Count != 0)
 									{
 										var lastIndex = Index[^1];
-										heightToRequest = lastIndex.Header.Height + 1;
+										currentHeight = lastIndex.Header.Height;
 										currentHash = lastIndex.Header.BlockHash;
 									}
+									else
+									{
+										currentHash = StartingHeight == 0 ? uint256.Zero :
+											await RpcClient.GetBlockHashAsync((int)StartingHeight - 1);
+										currentHeight = StartingHeight - 1;
+									}
 								}
 
-								// If not synchronized or already 5 min passed since last update, get the latest blockchain info.
-								if (!syncInfo.IsCoreSynchornized || DateTimeOffset.UtcNow - syncInfo.BlockchainInfoUpdated > TimeSpan.FromMinutes(5))
+								var coreNotSynced = !syncInfo.IsCoreSynchornized;
+								var tipReached = syncInfo.BlockCount == currentHeight;
+								var refresh = DateTimeOffset.UtcNow - syncInfo.BlockchainInfoUpdated > TimeSpan.FromMinutes(5);
+								if (coreNotSynced || tipReached || refresh)
```

Comment:
I like this change, it is much clear now.

nit:
```suggestion
								var isTimeToRefresh = DateTimeOffset.UtcNow - syncInfo.BlockchainInfoUpdated > TimeSpan.FromMinutes(5);
								if (coreNotSynced || tipReached || isTimeToRefresh)
```